[lldb/interpreter] Add Swift REPL init file support

medismailben opened the pull request:
Some users complained that when using the Swift REPL (LLDB launched with
`--repl` option), it would try to read their home init file, which can cause
unintended behaviour.

To solve this issue, this patch introduces a new, more specific, Swift
REPL init file, conveniently named `.lldbinit-swift-repl`. If the user
creates this file in his home directory, and launches LLDB in REPL mode,
it will override the default init file.

This behaviour only applies when REPL mode is enabled.

rdar://65836048

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

JDevlieghere committed:
[lldb] Provide GetHomeDirectory wrapper in Host::FileSystem (NFC)

Provider a wrapper around llvm::sys::path::home_directory in the
FileSystem class. This will make it possible for the reproducers to
intercept the call in a central place.

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test platform Linux

medismailben commented:
@swift-ci test linux platform

medismailben commented:
@swift-ci test

JDevlieghere commented:
@swift-ci test platform linux

JDevlieghere commented:
@swift-ci please test platform Linux

JDevlieghere commented:
@swift-ci please test linux

JDevlieghere commented on lldb/test/Shell/SwiftREPL/InitFile.test:
```diff
@@ -0,0 +1,15 @@
+// Test that the Swift REPL init file works.
+// REQUIRES: system-darwin
+
+// RUN: if [ -f ~/.lldbinit ]; then mv ~/.lldbinit ~/.lldbinit.old; fi
```
If the test fails it'll leave the lldbinit file in the old position. You can use `env HOME=%t` or something to spoof the home directory, we have other tests on llvm.org that do this too.

JDevlieghere commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -2098,7 +2099,9 @@ static void GetHomeREPLInitFile(llvm::SmallVectorImpl<char> &init_file,
   switch (language) {
-  // TODO: Add support for a language used with a REPL.
+  case eLanguageTypeSwift:
+    init_file_name = ".lldbinit-swift-repl";
```
Can we abstract this in the language plugin so that we don't need Swift specific logic in generic code?

JDevlieghere replied:
Another possible alternative is having a "scheme" where the REPL specific init file is always called `.lldbinit-<language>-repl`.

JDevlieghere reviewed (dismissed):
How are users supposed to know what file LLDB will read for the Swift REPL. You probably want to mention this in the man page. If you decide to go with the `.lldbinit-language-repl` scheme you could even do this on llvm.org.

medismailben commented:
@swift-ci test

medismailben commented:
I'll move this to llvm.org, and cherrypick it afterwards.

medismailben committed:
[lldb/interpreter] Improve REPL init file compatibility

This patch changes the command interpreter sourcing logic for the REPL
init file. Instead of looking for a arbitrary file name, it standardizes
the REPL init file name to match to following scheme:

                          `.lldbinit-<language>-repl`

This will make the naming more homogenous and the sourcing logic future-proof.

rdar://65836048

Differential Revision: https://reviews.llvm.org/D86987

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben committed:
[lldb/Interpreter] Fix language detection for the REPL InitFile

Previously, before loading the REPL language-specific init file, lldb
checked the selected target language in which case it returned an unknown
language type with the REPL target.

Instead, the patch calls `Language::GetLanguagesSupportingREPLs` and
look for the first element of that set. In case lldb was not configured
with a REPL language, then, it will just stop sourcing the REPL init
file and fallback to the original logic (continuing with the default
init file).

rdar://65836048

Differential Revision: https://reviews.llvm.org/D87076

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben committed:
[lldb/Interpreter] Add test for SwiftREPL initfile

This test adds `breakpoint set` commands with different source locations
to a temporary `.lldbinit-swift-repl` and `.lldbinit` and after launching
lldb, it checks that the breakpoint set in the debugger matches the
source location from the SwiftREPL initfile.

rdar://65836048

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

JDevlieghere commented:
Did you also cherry-pick the llvm.org patches to the Bastille branch?

medismailben commented:
> Did you also cherry-pick the llvm.org patches to the Bastille branch?

Yep, in #1736 and #1739 !
